tests/src/rules/no-cycle.js

```diff
@@ -178,69 +178,69 @@ ruleTester.run('no-cycle', rule, {
       test({
         code: `import { foo } from "./${testDialect}/depth-two"`,
         options: [{ ...opts }],
-        errors: [error(`Dependency cycle via ./depth-one:1`)],
+        errors: [error(`Dependency cycle detected.`)],
       }),
       test({
         code: `import { foo } from "./${testDialect}/depth-two"`,
         options: [{ ...opts, maxDepth: 2 }],
-        errors: [error(`Dependency cycle via ./depth-one:1`)],
+        errors: [error(`Dependency cycle detected.`)],
       }),
       test({
         code: `const { foo } = require("./${testDialect}/depth-two")`,
-        errors: [error(`Dependency cycle via ./depth-one:1`)],
+        errors: [error(`Dependency cycle detected.`)],
         options: [{ ...opts, commonjs: true }],
       }),
       test({
         code: `import { two } from "./${testDialect}/depth-three-star"`,
         options: [{ ...opts }],
-        errors: [error(`Dependency cycle via ./depth-two:1=>./depth-one:1`)],
+        errors: [error(`Dependency cycle detected.`)],
       }),
       test({
         code: `import one, { two, three } from "./${testDialect}/depth-three-star"`,
         options: [{ ...opts }],
-        errors: [error(`Dependency cycle via ./depth-two:1=>./depth-one:1`)],
+        errors: [error(`Dependency cycle detected.`)],
       }),
       test({
         code: `import { bar } from "./${testDialect}/depth-three-indirect"`,
         options: [{ ...opts }],
-        errors: [error(`Dependency cycle via ./depth-two:1=>./depth-one:1`)],
+        errors: [error(`Dependency cycle detected.`)],
       }),
       test({
         code: `import { bar } from "./${testDialect}/depth-three-indirect"`,
         options: [{ ...opts }],
-        errors: [error(`Dependency cycle via ./depth-two:1=>./depth-one:1`)],
+        errors: [error(`Dependency cycle detected.`)],
         parser: parsers.BABEL_OLD,
       }),
       test({
         code: `import { foo } from "./${testDialect}/depth-two"`,
         options: [{ ...opts, maxDepth: Infinity }],
-        errors: [error(`Dependency cycle via ./depth-one:1`)],
+        errors: [error(`Dependency cycle detected.`)],
       }),
       test({
         code: `import { foo } from "./${testDialect}/depth-two"`,
         options: [{ ...opts, maxDepth: '∞' }],
-        errors: [error(`Dependency cycle via ./depth-one:1`)],
+        errors: [error(`Dependency cycle detected.`)],
       }),
     ]).concat([
       test({
         code: `import("./${testDialect}/depth-three-star")`,
-        errors: [error(`Dependency cycle via ./depth-two:1=>./depth-one:1`)],
+        errors: [error(`Dependency cycle detected.`)],
         parser: parsers.BABEL_OLD,
       }),
       test({
         code: `import("./${testDialect}/depth-three-indirect")`,
-        errors: [error(`Dependency cycle via ./depth-two:1=>./depth-one:1`)],
+        errors: [error(`Dependency cycle detected.`)],
```

Comment:
hm, this seems like a pretty big expressivity loss in the error messages. is there no way to get this info from the SCC graph?

Reply:
The reported path is the shortest. We don't have that in the SCC.
In the SCC we have all of the files that have a path to each other, which may contain multiple cycles.

To remain backwards compatible, we need to run BFS to find the shortest cycle. That will have a performance cost, but we're still going to be more efficient than without SCC, since we only need to do this BFS if we know that there's a cycle. No BFS if there's no cycle.

Reply:
Could we just pick one of the paths? or, include them all in the message?

Reply:
"Different from the current message" is fine, if it's better :-)

The particular case I'm thinking about is when the cycle is a much, much larger circle - when a seemingly innocuous change ends up with a cycle far away from you, and it's good if the message can point you to the source of the problem.

Reply:
I think the best UX will be to keep the shortest path in the error message as it was, though it will have a perf cost due to the BFS as discussed.

Wdyt about adding a new config option to let the user decide if they prefer expressivity/precision vs speed?

Reply:
I suppose we could do that - but i'm not sure the performance cost of a failure case matters? when there's no cycles, it would be fast - it'd only be when there is a cycle that there'd be a cost to compute the shortest path

Reply:
Here I add a new option to skip the cycle-route BFS:

- https://github.com/import-js/eslint-plugin-import/pull/2998/commits/ebc2e80f4dd45d4f2592fede491bcf173014e093

Feel free to propose a better name.

Note that this lacks test coverage, and when on will definitely report cycles that it wouldn't when off.

Reply:
maybe i'm confused.

I'm saying that we should always do the fast approach. Then, only if there's a cycle detected, we can do the slow thing to come up with a helpful error message.

Reply:
Right. The way I measured the 323% increase is:

`"rules": ["import/no-cycle", { noRoute: true  }]` -> 1152 ms
`"rules": ["import/no-cycle", { noRoute: false }]` -> 4876 ms

Reply:
but in the case where there are no cycles, the perf is the same, yes?

Reply:
Yes, when there is no cycle, we don't have any additional work to do, so we take the exact same code path (early exit at https://github.com/import-js/eslint-plugin-import/pull/2998/files#diff-4b652cd5e29e44f86a5f087e0f8eec94bd9b6ef723d86dccd60656a044899bccR116).